storage/innobase/include/buf0buf.h

```diff
@@ -657,6 +657,22 @@ class buf_page_t
     access_time= 0;
   }
 
+private:
+  void set_memory_state(uint32_t s)
+  {
+#ifdef MADV_FREE
+    if (s == buf_page_t::NOT_USED)
+      madvise(frame, srv_page_size, MADV_FREE);
+#endif
+#ifdef _WIN32
+    if (s == buf_page_t::NOT_USED)
+      VirtualAlloc(frame, srv_page_size, MEM_RESET, PAGE_NOACCESS);
+    else if (state() == buf_page_t::NOT_USED)
+      VirtualAlloc(frame, srv_page_size, MEM_RESET_UNDO, PAGE_NOACCESS);
```

Comment:
This may introduce a performance regression on Windows, because `zip.fix` would have to be read each time before it is being assigned.

I think that a better approach would be to add new member functions for explicitly marking memory as unused or used wherever it is needed: `buf_block_init()`, `buf_LRU_block_free_non_file_page()`, `buf_pool_t::realloc()`, `buf_page_create_low()`, `buf_LRU_get_free_only()`.

The `NOT_USED` state is also being used in `buf_pool_t::watch_set()` on pseudo block descriptors that do not have any page frame attached to them. We should not invoke any system calls in that case. Perhaps you did not test with the non-default parameter `innodb_change_buffering=all` or the test was too small to exercise the change buffer.

By the way, there is no need to qualify the `enum` constants with `buf_page_t::` in a member function of `buf_page_t`.

Reply:
`TRANSACTIONAL_TARGET` is only something to make GCC happy. Any function that potentially uses memory transactions (directly, or by invoking functions declared with `TRANSACTIONAL_INLINE`) must declare that attribute. Any use of memory transaction instructions should be guarded by checking the flag `have_transactional_memory`.

All memory transactions are short, and we never execute system calls within a memory transaction.

Did you confuse `buf_page_create_low()` and `buf_page_init_low()`? The latter is invoked by `buf_block_t::initialise()` and `buf_page_get_low()`. Those do not use `TRANSACTIONAL_TARGET`, but `buf_page_create_low()` does, for adding the page to `buf_pool.page_hash`:
```c++
  bpage->lock.x_lock();

  /* The block must be put to the LRU list */
  buf_LRU_add_block(bpage, false);
  {
    transactional_lock_guard<page_hash_latch> g
      {buf_pool.page_hash.lock_get(chain)};
    bpage->set_state(buf_page_t::REINIT + 1);
    buf_pool.page_hash.append(chain, bpage);
  }
```
Any time after this, we may invoke the system call to note that `bpage->frame` is being used. The frame will be protected by the exclusive `bpage->lock` that this thread will be holding.